pkg/agent/controller/networkpolicy/cache.go

```diff
@@ -140,16 +143,40 @@ func policyIndexFunc(obj interface{}) ([]string, error) {
 }
 
 // newRuleCache returns a new *ruleCache.
-func newRuleCache(dirtyRuleHandler func(string)) *ruleCache {
+func newRuleCache(dirtyRuleHandler func(string), podUpdate <-chan v1beta1.PodReference) *ruleCache {
 	rules := cache.NewIndexer(
 		ruleKeyFunc,
 		cache.Indexers{addressGroupIndex: addressGroupIndexFunc, appliedToGroupIndex: appliedToGroupIndexFunc, policyIndex: policyIndexFunc},
 	)
-	return &ruleCache{
+	cache := &ruleCache{
 		podSetByGroup:     make(map[string]podSet),
 		addressSetByGroup: make(map[string]sets.String),
 		rules:             rules,
 		dirtyRuleHandler:  dirtyRuleHandler,
+		podUpdates:        podUpdate,
+	}
+	go cache.processPodUpdates()
+	return cache
+}
+
+// processPodUpdates is an infinite loop that takes Pod update events from the
+// channel, finds out AppliedToGroups that contains this Pod and trigger
+// reconciling of related rules.
+func (c *ruleCache) processPodUpdates() {
+	for {
+		select {
+		case pod := <-c.podUpdates:
+			func() {
+				c.podSetLock.RLock()
+				defer c.podSetLock.RUnlock()
+				for group, podSet := range c.podSetByGroup {
+					_, exists := podSet[pod]
+					if exists {
+						c.onAppliedToGroupUpdate(group)
```

Comment:
Maybe a stupid question. Are we assuming that by the time the CNI server notifies us the channel and we execute this code, we have already received the created / updated relevant AppliedToGroups from the Antrea NP controller? If yes, it may be worth documenting in the comment for this function (and explaining why that is the case). If no, then I think I would benefit from an explanation, because in my mind the sequence of events is as follows:
1) Antrea NP controller is notified by apiserver that a new Pod is being created. The IP address is not known yet, but we update / create AppliedToGroups as needed based on existing Network Policies. 
2) Agent NP controller is notified by Antrea NP controller, flows cannot be installed yet because IP address is not known.
3) CNIAdd is invoked, Pod networking is configured, IP address is determined. CNI server can notify Agent NP controller using the `podUpdates` channel.
4) `processPodUpdates` processes the notification from the CNI server. The IP address is now retrievable from the interface store, and we can install NP flows.

So I guess my question is: is this the correct sequence of events, are we assuming that 2) happens before 3) (and if yes why can we make this assumption)?

Reply:
The NetworkPolicy implementation is all asynchronous from Antrea Controller to Antrea Agent with best-effort. Your understanding of the sequence of events is correct, but we don't assume it must happen,  just it should be that case with a great possibility.
It's because the way Kubernetes works: kube-scheduler schedules a Pod on a Node by setting its its `NodeName` field via API, then kubelet and antrea-controller can receive the Pod update event at almost the same time, and kubelet starts to create containers for the Pod and antrea-controller starts to update the Group info and notify antrea-agent. It's observed that the former is slower than the latter so the PR works in that case. But it's hard to say it's consistent, due to different network condition, antrea-controller workload, kubelet workload, and cluster scale.
This PR is to enhance the best-effort by triggering the realization as early as possible. Even if a NetworkPolicy is created before a Pod, we can't say antrea-controller must receive the Policy from APIServer first as they are two API connections. But this slight difference should make no difference to users as they should reach eventual consistency in a very short time.
I will add comments to explain which situation this processing will help.